notes-wasm.md

```diff
@@ -22,10 +22,157 @@ cargo test --release --test poseidon_wasm --no-run --target wasm32-unknown-unkno
 cargo build --release --target wasm32-unknown-unknown
 ```
 
-# Build and review the wat
+# Build and review the output
 
+## Review the wasm assembly
 From the folder `wasm`:
 ```
 cargo build --release --target wasm32-unknown-unknown
 wasm2wat ../target/wasm32-unknown-unknown/release/wasm_test.wasm -o wasm_test.wat
-``
+```
+
+Then in `wasm_test.wat` find the relevant function.
+
+## Review the code generated by Spidermonkey (the Firefox js engine)
+
+Download jsshell from https://archive.mozilla.org/pub/firefox/nightly/latest-mozilla-central/jsshell-linux-x86_64.zip
+Start it and run the following lines:
+```
+m = new WebAssembly.Module(os.file.readFile("add.wasm","binary"));
+c = wasmExtractCode(m);
+os.file.writeTypedArrayToFile('out.bin', c.code);
+```
+Then disassembley the result using radare2
+```
+rasm2 -S intel -a x86 -b 64 -D -B -f out.bin
+```
+
+## Native compilation generated code
+```
+cargo build --release
+objdump -d ../target/release/libwasm_test.so
+```
+Then in the disassembly output find the relevant function.
+
+## `GoldilocksField Add`
+
+### Rust source
+```rust
+fn add(self, rhs: Self) -> Self {
+    let (sum, over) = self.0.overflowing_add(rhs.0);
+    let (mut sum, over) = sum.overflowing_add((over as u64) * EPSILON);
+    if over {
+        // NB: self.0 > Self::ORDER && rhs.0 > Self::ORDER is necessary but not sufficient for
+        // double-overflow.
+        // This assume does two things:
+        //  1. If compiler knows that either self.0 or rhs.0 <= ORDER, then it can skip this
+        //     check.
+        //  2. Hints to the compiler how rare this double-overflow is (thus handled better with
+        //     a branch).
+        assume(self.0 > Self::ORDER && rhs.0 > Self::ORDER);
+        branch_hint();
+        sum += EPSILON; // Cannot overflow.
+    }
+    Self(sum)
+}
+```
+
+### WASM
+```wat
+  (func $test_goldilocks_add (type 15)
+    (local i64 i64)
+    call $get_input ;; [a]
+    local.tee 0 ;; $0 = a [a]
+    call $get_input ;; [a, b]
+    i64.add ;; sum0 = a+b [sum0]
+    local.tee 1 ;; $1 = sum0;
+    i64.const 4294967295 ;; [sum0, 0xffffffff]
+    i64.const 0 ;; [sum0, 0xffffffff, 0]
+    local.get 1 ;; [sum0, 0xffffffff, 0, sum0]
+    local.get 0 ;; [sum0, 0xffffffff, 0, sum0, a]
+    i64.lt_u ;; over0 = sum0 < a [sum0, 0xffffffff, 0, over0]
+    select ;; [sum0, over0 ? 0xffffffff : 0]
+    i64.add ;; sum1 = sum0 + (over0 ? 0xffffffff : 0) [sum1]
+    local.tee 0 ;; $0 = sum1 [sum1]
+    i64.const 4294967295 ;; [sum1, 0xffffffff]
+    i64.add ;; [sum1 + 0xffffffff]
+    local.get 0 ;; [sum1 + 0xffffffff, sum1]
+    local.get 0 ;; [sum1 + 0xffffffff, sum1, sum1]
+    local.get 1 ;; [sum1 + 0xffffffff, sum1, sum1, sum0]
+    i64.lt_u ;; over1 = sum1 < sum0 [sum1 + 0xffffffff, sum1, over1]
+    select ;; sum2 = over1 ? sum1 + 0xffffffff ; sum1 [sum2]
+    call $set_output)
```

Comment:
I've just realized that this is missing the last 8 lines, if the code stops here it returns a wrong result

the missing lines:
```wat
    local.tee 0
    i64.const 4294967295
    i64.add
    local.get 0
    local.get 0
    i64.const -4294967296
    i64.gt_u
    select
```

Reply:
I think the part that you find missing corresponds to the canonicalization function.  Without it the result is correct but not canonical, which means that sometimes you'd get `k+p` instead of `k` (as long as `k+p < 2^64`).
But since the field operations can be made on the non-canonical form I removed it so that we can see the amortized cost (because the canonicalization would only be done once after all the arithmetic operations)

Reply:
can you check if subtracting `p` from the result yields the value you expect?  where `p = 2^64 - 2^32 + 1`